File: 백준/Gold/34409. 무등산 등반/무등산 등반.cc
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
using namespace std;
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);

	int N, M; cin >> N >> M;
	int ix, iy; cin >> ix >> iy;
	ix--; iy--;
	int a, b, c; cin >> a >> b >> c;
	vector<vector<int>>H(N, vector<int>(M));

	int tpx = -1, tpy = -1, tph = 0;
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			int ih; cin >> ih;
			H[i][j] = ih;
			if (tph < H[i][j]) {
				tph = H[i][j];
				tpx = i; tpy = j;
			}
		}
	}

	const long long INF = 5e9;
	int dx[4] = { 1,-1,0,0 };
	int dy[4] = { 0,0,1,-1 };
	vector<vector<long long>> dist(N, vector<long long>(M, INF));
	using Node = tuple<long long, int, int>;
	priority_queue<Node, vector<Node>, greater<Node>> pq;

	dist[ix][iy] = 0;
	pq.emplace(0, ix, iy);

	while (!pq.empty()) {
		Node cur = pq.top();
		long long d = get<0>(cur);
		int cx = get<1>(cur);
		int cy = get<2>(cur);
		pq.pop();
		
		if (d != dist[cx][cy]) continue;

		if (cx == tpx && cy == tpy) {
			cout << d << "\n";
			return 0;
		}

		for (int i = 0; i < 4; i++) {
			int nx = cx + dx[i];
			int ny = cy + dy[i];
			if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
			int dh = H[nx][ny] - H[cx][cy];
			if (abs(dh) > c) continue;

			long long w = (dh == 0) ? 1 : ((dh > 0) ? 1LL * dh * a : 1LL * (-dh) * b);
			
			if (dist[nx][ny] > d + w) {
				dist[nx][ny] = d + w;
				pq.emplace(dist[nx][ny], nx, ny);
			}
		}
	}
	cout << -1 << "\n";
	return 0;

}

File: 백준/Gold/34409. 무등산 등반/무등산 등반.py
import sys
import heapq


input = sys.stdin.readline

if __name__ == "__main__":
    N,M = map(int, input().split())
    ix, iy = map(int, input().split())
    ix, iy = ix - 1, iy - 1
    a, b, c = map(int, input().split())
    H = []

    tpx, tpy, tph = -1, -1, -1
    for i in range(N):
        row = list(map(int,input().split()))
        H.append(row)
        for j, h in enumerate(row):
            if h > tph:
                tph = h
                tpx, tpy = i, j

    INF = int(5e9)
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]
    dist = [[INF for j in range(M)] for i in range(N)]
    pq = []
    dist[ix][iy] = 0

    heapq.heappush(pq, (0, ix, iy))

    while pq:
        d, x, y = heapq.heappop(pq)
        
        if d != dist[x][y]:
            continue

        if x == tpx and y == tpy:
            print(d)
            sys.exit(0)

        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if nx < 0 or ny < 0 or nx >= N or ny >= M:
                continue
            dh = H[nx][ny] - H[x][y]
            if abs(dh) > c:
                continue

            w = 1 if dh == 0 else (dh * a if dh > 0 else -dh * b)

            if dist[nx][ny] > d + w:
                dist[nx][ny] = d + w
                heapq.heappush(pq, (dist[nx][ny], nx, ny))
    
    print(-1)




